File: src/lib/analysis/concentration.ts
import type { Trade } from "@/types/database";
import type { Position } from "@/lib/portfolio";

export const HHI_HIGH = 0.5;          // 집중 — 경고 수준
export const HHI_MID = 0.25;          // 보통
export const TOP1_WEIGHT_HIGH = 0.4;  // 단일 종목 비중 경고 임계치

export interface ConcentrationData {
  hhi: number;
  top3: { asset: string; weight: number }[];
  byCountry: { code: string; weight: number }[];
  byMarket: { type: string; weight: number }[];
}

export function computeConcentration(positions: Position[], trades: Trade[]): ConcentrationData {
  // evaluation이 없으면 costBasis로 fallback
  const values = positions.map((p) => ({
    key: p.key,
    asset: p.assetName,
    country: p.country,
    value: p.evaluation ?? p.costBasis,
  }));

  const total = values.reduce((s, v) => s + v.value, 0);

  if (total === 0) {
    return { hhi: 0, top3: [], byCountry: [], byMarket: [] };
  }

  // HHI: Σ(weight²)
  const hhi = values.reduce((s, v) => {
    const w = v.value / total;
    return s + w * w;
  }, 0);

  // top3
  const top3 = [...values]
    .sort((a, b) => b.value - a.value)
    .slice(0, 3)
    .map((v) => ({ asset: v.asset, weight: v.value / total }));

  // byCountry
  const countryMap = new Map<string, number>();
  for (const v of values) {
    countryMap.set(v.country, (countryMap.get(v.country) ?? 0) + v.value);
  }
  const byCountry = Array.from(countryMap.entries())
    .map(([code, val]) => ({ code, weight: val / total }))
    .sort((a, b) => b.weight - a.weight);

  // byMarket: trades에서 종목별 market_type 추출 (가장 최근 BUY 기준)
  const marketByKey = new Map<string, string>();
  for (const t of trades
    .filter((t) => t.trade_type === "BUY")
    .sort((a, b) => new Date(a.traded_at).getTime() - new Date(b.traded_at).getTime())) {
    const key = `${t.ticker_symbol ?? t.asset_name}:${t.country_code ?? "KR"}`;
    marketByKey.set(key, t.market_type);
  }

  const marketMap = new Map<string, number>();
  for (const v of values) {
    const mt = marketByKey.get(v.key) ?? "ETC";
    marketMap.set(mt, (marketMap.get(mt) ?? 0) + v.value);
  }
  const byMarket = Array.from(marketMap.entries())
    .map(([type, val]) => ({ type, weight: val / total }))
    .sort((a, b) => b.weight - a.weight);

  return { hhi, top3, byCountry, byMarket };
}

File: api/src/invest_note_api/domain/analysis/concentration.py
"""집중도(HHI) 계산."""
from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from invest_note_api.domain.portfolio import Position
    from invest_note_api.domain.trade_types import Trade

HHI_HIGH = 0.5
HHI_MID = 0.25
TOP1_WEIGHT_HIGH = 0.4


@dataclass
class ConcentrationData:
    hhi: float
    top3: list[dict]          # [{"asset": str, "weight": float}]
    by_country: list[dict]    # [{"code": str, "weight": float}]
    by_market: list[dict]     # [{"type": str, "weight": float}]


def compute_concentration(positions: list[Position], trades: list[Trade]) -> ConcentrationData:
    values = [
        {
            "key": p.key,
            "asset": p.asset_name,
            "country": p.country,
            "value": p.evaluation if p.evaluation is not None else p.cost_basis,
        }
        for p in positions
    ]

    total = sum(v["value"] for v in values)
    if total == 0:
        return ConcentrationData(hhi=0.0, top3=[], by_country=[], by_market=[])

    hhi = sum((v["value"] / total) ** 2 for v in values)

    top3 = [
        {"asset": v["asset"], "weight": v["value"] / total}
        for v in sorted(values, key=lambda x: x["value"], reverse=True)[:3]
    ]

    country_map: dict[str, float] = {}
    for v in values:
        country_map[v["country"]] = country_map.get(v["country"], 0.0) + v["value"]
    by_country = sorted(
        [{"code": k, "weight": val / total} for k, val in country_map.items()],
        key=lambda x: x["weight"],
        reverse=True,
    )

    market_by_key: dict[str, str] = {}
    for t in sorted(trades, key=lambda t: t.traded_at):
        if t.trade_type == "BUY":
            key = f"{t.ticker_symbol or t.asset_name}:{t.country_code or 'KR'}"
            market_by_key[key] = t.market_type

    market_map: dict[str, float] = {}
    for v in values:
        mt = market_by_key.get(v["key"], "ETC")
        market_map[mt] = market_map.get(mt, 0.0) + v["value"]
    by_market = sorted(
        [{"type": k, "weight": val / total} for k, val in market_map.items()],
        key=lambda x: x["weight"],
        reverse=True,
    )

    return ConcentrationData(hhi=hhi, top3=top3, by_country=by_country, by_market=by_market)
